File: T-Bank/Algorithms-and-Data-Structures-2025/12sprint-SolvingProblemMethodsPart2/contest/F-Монетки/solution.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;


void solution() {
    /*
    Time Complexity O(2^M)
    Memory Complexity O(M)
    */
    int N, M;
    scanf("%d %d", &N, &M);
    vector<int> coins(M);
    for (int i = 0; i < M; ++i) scanf("%d", &coins[i]);
    int min_coins = INT_MAX;
    vector<int> best_combination;
    for (int mask = 0; mask < (1 << (2 * M)); ++mask) {
        int total = 0;
        int count = 0;
        vector<int> current;
        bool valid = true;
        for (int i = 0; i < M; ++i) {
            int bits = (mask >> (2 * i)) & 0x3;
            if (bits > 2) {
                valid = false;
                break;
            }
            total += bits * coins[i];
            count += bits;
            for (int j = 0; j < bits; ++j) current.push_back(coins[i]);
        }
        if (valid && total == N) {
            if (count < min_coins) {
                min_coins = count;
                best_combination = current;
            }
        }
    }

    if (min_coins != INT_MAX) {
        printf("%d\n", min_coins);
        for (int coin : best_combination) printf("%d ", coin);
        printf("\n");
    } else {
        int totalMoney = 0;
        for (int coin : coins) totalMoney += 2 * coin;
        printf("%d\n", (totalMoney < N? -1 : 0));
    }
}


int main() {
    solution();
}

File: T-Bank/Algorithms-and-Data-Structures-2025/12sprint-SolvingProblemMethodsPart2/contest/F-Монетки/solution.py
import sys


def solution() -> None:
    """
    Time Complexity O(2**M)
    Memory Complexity O(M)
    """
    N, M = map(int ,sys.stdin.readline().rstrip().split())
    coins: list[int] = list(map(int, sys.stdin.readline().rstrip().split()))
    min_coins: int = float('inf')
    best_combination: list[int] = []
    for mask in range(1 << (2 * M)):
        total: int = 0
        count: int = 0
        current: list[int] = []
        valid: bool = True
        for i in range(M):
            bits: int = (mask >> (2 * i)) & 0x3
            if bits > 2:
                valid = False
                break 
            total += bits * coins[i]
            count += bits
            for j in range(bits): current.append(coins[i])
        if valid and total == N:
            if count < min_coins:
                min_coins = count
                best_combination = current
    if min_coins != float('inf'):
        sys.stdout.write('{}\n{}\n'.format(min_coins, ' '.join(str(coin) for coin in best_combination)))
    else:
        total_money: int = 0
        for coin in coins: total_money += 2 * coin
        sys.stdout.write('{}\n'.format([0, -1][total_money < N]))


if __name__ == '__main__':
    solution()